### PR TITLE
Disallow destructuring assignment in `using` declarations

### LINKED ISSUE #16604: [Bug]: `using` declarations shoud disallow destructuring
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Other (Next.js, Gatsby, vue-cli, ...)

### Input code

```js
using x = null, {} = null;
```
and
```js
using x = null, [] = null;
```

### Configuration file name

_No response_

### Configuration

Using `@babel/plugin-proposal-explicit-resource-management`

### Current and expected behavior

It should throw a parse error, but it doesn't

### Environment

Babel 7.24

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by H0onnn**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

- close #16604 

A syntax error occurs when using destructuring assignment in a using declaration.

### CODE REVIEW

**Review by nicolo-ribaudo — APPROVED** [review]
Thank you! This PR is already good as is, but it would be even better if we could give a good error also for this case:
```js
using { x } = y;
```
by adapting the logic in https://github.com/H0onnn/babel/blob/3ae03e1b53c69eec8850071eca53908c09c1d6ed/packages/babel-parser/src/parser/statement.ts#L505-L513 so that if `using` is followed on the same line by `{` (other than by an identifier) we parse it as an using declaration, and then it will raise the error that you added in this PR.

Unfortunately we cannot do the same for `[`, because `using [x] = y` is valid code (setting the `x` property of the variable `using`).

**Comment by H0onnn** [comment]
@nicolo-ribaudo Hello! Thank you for the quick review. Added code to allow for 'ArrayPattern'. Will this work meet your requirements?

**Comment by nicolo-ribaudo** [comment]
@H0onnn No, your original commit was correct :) We must report an error for `using y = null, [] = null;`.

My suggestion was to change the logic in https://github.com/H0onnn/babel/blob/3ae03e1b53c69eec8850071eca53908c09c1d6ed/packages/babel-parser/src/parser/statement.ts#L505-L513 to also check for `{` after `using`, so that we also get the nice error message for the `using { x } = y` case.

**Comment by H0onnn** [comment]
@nicolo-ribaudo I understand. Please understand that my English skills are not good. 😅

I'll keep trying !

File: packages/babel-parser/test/fixtures/placeholders/variable/using-init/output.json [hunk 1]
```diff
@@ -1,6 +1,9 @@
 {
   "type": "File",
   "start":0,"end":30,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":3,"column":1,"index":30}},
+  "errors": [
+    "SyntaxError: Using declaration cannot have destructuring patterns. (2:8)"
+  ],
```
Comment (JLHwung) [inline packages/babel-parser/test/fixtures/placeholders/variable/using-init/output.json:6]: This is a valid case, when the placeholder plugin is enabled, `using %%name%% = null` is a valid production. Can we throw the error when `id` is either `ArrayPattern` or `ObjectPattern`?

Comment (H0onnn) [reply packages/babel-parser/test/fixtures/placeholders/variable/using-init/output.json]: Hello @JLHwung! Thank you for your feedback. As you said, modified it to throw an error only in the case of ObjectPattern and ArrayPattern.

**Comment by H0onnn** [comment]
@nicolo-ribaudo I hope that the logic added this commit will satisfy your requirements!

**Review by nicolo-ribaudo — APPROVED** [review]
I pushed a commit to improve the error message of `using {` (and added a test) -- thank you!

**Review by JLHwung — APPROVED** [review]
Thanks.

**Pull request merged**